### PR TITLE
Add stream compress for gpdb6

### PR DESCRIPTION
**Pull request by Hlinbit**
Gpfdist is a high-performance ETL tool to load external data for gpdb. In practice,  while gpfdist can extract the full performance of the network card, gpfdist can also affect other processes on the same host.

Especially in cloud services, multiple services may share one physical network card. If Gpfdist exclusively uses the network card, it will lead to the failure of other services.

If both high transmission efficiency and low network usage are required, transferring the data after compression is a reasonable approach. Thus this code change is about using stream compression based on zstd algorithm to transfer data from gpfdist to gpdb. The code change involves the following aspects:

-  The switch flag(ec) to turn on/off compression transmission for gpfdist.
-  The modification of the HTTP header for requests and responses for interaction between gpdb and gpfdist.
-  The modification of compression data(for gpfdist end) and decompressing data(for gpdb end) streamingly.

Finally, to use the zstd compression to transfer data, you can add the specific flag `--ec` when starting up gpfdist. For example, typing `gpfdist -d /home/gpadmin -p 7070 --ec`  to start gpfdist, and then data would be compressed from gpfdist to gpdb.

### CODE REVIEW

File: src/backend/access/external/url_curl.c [hunk 1]
```diff
@@ -347,6 +352,39 @@ header_callback(void *ptr_, size_t size, size_t nmemb, void *userp)
 		}
 	}
 
+	if (len > 9 && *ptr == 'X' && 0 == strncmp("X-GP-ZSTD", ptr, 9))
+	{
+		ptr += 9;
+		len -= 9;
```
Comment (flykos) [inline src/backend/access/external/url_curl.c:358]: what is 9 ?

File: src/backend/access/external/url_curl.c [hunk 2]
```diff
@@ -1514,6 +1554,27 @@ gp_proto0_read(char *buf, int bufsz, URL_CURL_FILE *file)
...
+		* be multiple frames concatenated together. Zstd will automatically
+		* reset the context when a frame is complete. Still, calling
+		* ZSTD_DCtx_reset() can be useful to reset the context to a clean
+		* state, for instance if the last decompression call returned an
+		* error.
+		*/
+		 
+	ret = ZSTD_decompressStream(ctx, bout, bin);
+	size_t const err = ret;                          
+	if(ZSTD_isError(err)){
```
Comment (hyongtao-db) [inline src/backend/access/external/url_curl.c:1570]: The recommended format is

```
if (X) 
{
	return A;
}
else
{
	return B;
}
```

File: src/bin/gpfdist/gpfdist.c [hunk 3]
```diff
@@ -172,7 +173,8 @@ static struct
 	struct transform* trlist; /* transforms from config file */
 	const char* ssl; /* path to certificates in case we use gpfdist with ssl */
 	int			w; /* The time used for session timeout in seconds */
-} opt = { 8080, 8080, 0, 0, 0, ".", 0, 0, -1, 5, 0, 32768, 0, 256, 0, 0, 0, 0 };
+	int			compress;
```
Comment (hyongtao-db) [inline src/bin/gpfdist/gpfdist.c:176]: Is `boolean` more suitable for the variable` compress`?

**Comment by kainwen** [comment]
I see this PR is for 6X and is under review, do we have 7X version? Most of the time, the workflow is

1. start PR to master (better, for new feature, have a detailed discussion on dev mailing list first)
2. after solid review, test the master branch PR pushed, then let's open a backport PR to previous version and register the backport PR in ship-it

The only condition that we start from previous version is that the code is very different from master version. But for gpfdist, obviously, it is not this case.

---------

Besides, I see a PR https://github.com/greenplum-db/gpdb/pull/13983 to master branch that PR conains a detailed performance test (on 6X). I play a bit, it can be cherry-picked to 6X very smoothly.

So why this PR?

File: src/backend/access/external/url_curl.c (hunk 1 above)
Comment (Hlinbit) [reply src/backend/access/external/url_curl.c:358]: It is the lenght of 'X-GP-ZSTD'.

File: src/backend/access/external/url_curl.c [hunk 4]
```diff
@@ -1857,6 +1932,18 @@ url_curl_fflush(URL_FILE *file, CopyState pstate)
 	gp_proto0_write((URL_CURL_FILE *) file, pstate);
 }
 
+size_t
+decompress_data(char* bufin, int len){
+	unsigned long long const rSize = ZSTD_getFrameContentSize(bufin, len);
+	void* const rBuff = malloc(rSize);
+    if (!rBuff) {
+		elog(ERROR, "decompress_data error: there is no enough space, required size is %llu", rSize);
```
Comment (hyongtao-db) [inline src/backend/access/external/url_curl.c:1940]: Missing unit 'Bytes': `elog(ERROR, "decompress_data error: there is no enough space, required size is %llu Bytes.", rSize);`

**Comment by water32** [comment]
Sequential compression's performance is lower than non compression. Please refer to my PR using parallel asynchronous compression:
https://github.com/greenplum-db/gpdb/pull/13983

**Comment by Hlinbit** [comment]
> Sequential compression's performance is lower than non compression. Please refer to my PR using parallel asynchronous compression: #13983



> Sequential compression's performance is lower than non compression. Please refer to my PR using parallel asynchronous compression: #13983

**Comment by lij55** [comment]
> I see this PR is for 6X and is under review, do we have 7X version? Most of the time, the workflow is
> 
> 1. start PR to master (better, for new feature, have a detailed discussion on dev mailing list first)
> 2. after solid review, test the master branch PR pushed, then let's open a backport PR to previous version and register the backport PR in ship-it
> 
> The only condition that we start from previous version is that the code is very different from master version. But for gpfdist, obviously, it is not this case.
> 
> Besides, I see a PR #13983 to master branch that PR conains a detailed performance test (on 6X). I play a bit, it can be cherry-picked to 6X very smoothly.
> 
> So why this PR?

#13983 is not a fully working PR that can lead to that performance improvement. multiple process gpfdist will be a scope for GP7 only. such a big change is not suitable for stable version.
transmission compression is a good feature for both GP6 and GP7. GP7 compression transmission should based on multiple process enhancement based on @water32's existing work.  This PR does not focus on performance improvement but bandwidth save

the same PR for GP7 is unnecessary  this is also why this PR is for GP6 only as well.

**Review by zhaorui20005 — COMMENTED** [review]
Also I suggest to tidy the large number of commits into some important commits by "git rebase" command.

**Comment by huiliang-liu** [comment]
Please remember to update the PR description due to the flag changed

**Pull request closed without merging**